pysap/plugins/mri/reconstruct_3D/fourier.py

```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# XXX - Copyright (C) XXX, 3017
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+Fourier operators for cartesian and non-cartesian space.
+"""
+
+
+# Package import
+import warnings
+from .utils import convert_locations_to_mask_3D
+from pysap.plugins.mri.reconstruct.fourier import FourierBase
+from modopt.interface.errors import warn
+
+# Third party import
+try:
+    from pynufft import NUFFT_hsa
+    from pynufft import NUFFT_cpu
+
+except Exception:
+    warnings.warn("pyNUFFT python package has not been found. Exists only "
+                  "in Python3.x. Try Pynfft if non uniform Fourier"
+                  " transform needed")
+
+import numpy as np
+
+
+class Singleton:
+    """  Singleton design pattern, to count the number of instanciation of
+    a subclass.
+
+    Attributes
+    ----------
+    numOfInstances: int
+        number of instances of the Singleton class.
+    """
+
+    numOfInstances = 0
+
+    def countInstances(cls):
+        cls.numOfInstances += 1
+
+    countInstances = classmethod(countInstances)
+
+    def getNumInstances(cls):
+        return cls.numOfInstances
+    getNumInstances = classmethod(getNumInstances)
+
+    def __init__(self):
+        self.countInstances()
```

Comment:
Why not using a metaclass? 

<pre>
from pisap.base.utils import with_metaclass

class Singleton(type):
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]

class NUFFT(with_metaclass(Singleton)):
    ...
</pre>